crates/spawner/src/destroyer.rs

```diff
@@ -1,33 +1,254 @@
+                .in_set(SpawnerSet::Destroyer)
+                .in_set(DestroyerSet::DestroyDead),
         );
     }
 }
 
-fn destroy(
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum DestroyerSet {
+    FindDead,
+    SendData,
+    DestroyDead
+}
+
+// Find all entities with low health and mark them for despawning
+fn find_dead(
     mut commands: Commands,
     mut counter: ResMut<ObjectCounter>,
     entities: Query<(Entity, &Player, &ObjectType, &Health), Changed<Health>>,
+    mut event_writer: EventWriter<Despawning>,
 ) {
     for (entity, &player, &object_type, health) in entities.iter() {
         if health.destroyed() {
             if let ObjectType::Active(active_type) = object_type {
                 counter.player_mut(player).unwrap().update(active_type, -1);
             }
             commands.entity(entity).despawn_recursive();
+            event_writer.send(Despawning(entity));
+        }
+    }
+}
+
+fn destroy(mut commands: Commands, mut despawning: EventReader<Despawning>,) {
+    for entity in despawning.iter() {
+        commands.entity(entity.0).despawn_recursive();
+    }
+}
+
+
+#[derive(Debug)]
+pub struct DespawnEventsPlugin<Q, T, F = ()>
+where
+    F: ReadOnlyWorldQuery + Send + Sync + 'static,
+    T: Send + Sync + 'static + Clone + Component,
+    Q: WorldQuery + Send + Sync + 'static,
+{
+    _marker: PhantomData<(Q, F, T)>,
+}
+
+impl<
+        Q: WorldQuery + Send + Sync + 'static,
+        T: Send + Sync + 'static + Clone + Component,
+        F: ReadOnlyWorldQuery + Send + Sync + 'static,
+    > Default for DespawnEventsPlugin<Q, T, F>
+{
+    fn default() -> Self {
+        Self {
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<
+        Q: WorldQuery + Send + Sync + 'static,
+        T: Send + Sync + Clone + Component + 'static,
+        F: ReadOnlyWorldQuery + Send + Sync + 'static,
+    > Plugin for DespawnEventsPlugin<Q, T, F>
+{
+    fn build(&self, app: &mut App) {
+        app.add_event::<DespawnedComponents<Q, T, F>>().add_system(
+            send_data::<Q, T, F>
+                .in_base_set(GameSet::Update)
+                .run_if(in_state(AppState::InGame))
+                .after(DestroyerSet::FindDead)
+                .in_set(SpawnerSet::Destroyer)
+                .in_set(DestroyerSet::SendData),
+        );
+    }
+}
+
+// Contains the entity and the data of the despawned component
+#[derive(Debug)]
+pub struct DespawnedComponents<Q, T, F = ()>
+where
+    F: ReadOnlyWorldQuery + Send + Sync + 'static,
+    T: Send + Sync + 'static + Clone + Component,
+    Q: WorldQuery + Send + Sync + 'static,
+{
+    pub entity: Entity,
+    pub data: T,
+    _filter_marker: PhantomData<(Q, F)>,
+}
+
+struct Despawning(Entity);
+
+#[allow(unused)]
+fn send_data<'w, Q, T, F>(
+    mut despawning: EventReader<Despawning>,
+    mut events: EventWriter<DespawnedComponents<Q, T, F>>,
+    data: Query<Q, F>,
+) where
+    T: Clone + Component + Send + Sync + 'w,
```

Comment:
I think that this should be changed to something like

```rust
fn send_data<'w, Q, F, T>(
    mut despawning: EventReader<DespawnEvent>,
    mut events: EventWriter<DespawnedEvent<<<<Q as WorldQuery>::ReadOnly as WorldQuery>::Item<'w> as ToOwned>::Owned>>,
    data: Query<Q, F>,
) where
    Q: WorldQuery + Send + Sync + 'w,
    F: ReadOnlyWorldQuery + Send + Sync + 'w,
    <<Q as WorldQuery>::ReadOnly as WorldQuery>::Item<'w>: Clone + Send + Sync,
```

Reply:
I don't think this is possible. I gave it my best shot but I think the current implementation is as close as we will get. feel free to have a go on your own.